tests/Microsoft.Identity.Web.Test.Integration/AcquireTokenForAppIntegrationTests.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Web.Test.Common;
+using Microsoft.Identity.Web.Test.Common.Mocks;
+using Microsoft.Identity.Web.Test.Common.TestHelpers;
+using Microsoft.Identity.Web.Test.LabInfrastructure;
+using Microsoft.Identity.Web.TokenCacheProviders.InMemory;
+using System;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Identity.Web.Test.Integration
+{
+    public class AcquireTokenForAppIntegrationTests
+    {
+        TokenAcquisition _tokenAcquisition;
+        ServiceProvider _provider;
+        private MsalTestTokenCacheProvider _msalTestTokenCacheProvider;
+
+        private KeyVaultSecretsProvider _keyVault;
+        private string _ccaSecret;
+        private readonly ITestOutputHelper _output;
+
+        public AcquireTokenForAppIntegrationTests(ITestOutputHelper output) //test set-up
+        {
+            _output = output;
+
+            _keyVault = new KeyVaultSecretsProvider();
+            _ccaSecret = _keyVault.GetSecret(TestConstants.ConfidentialClientKeyVaultUri).Value;
+
+            if (!string.IsNullOrEmpty(_ccaSecret)) //Need the secret before building the services
+            {
+                BuildTheRequiredServices();
+            }
+            else
+            {
+                _output.WriteLine("Connection to keyvault failed. No secret returned. ");
+                throw new ArgumentNullException(nameof(_ccaSecret), "No secret returned from keyvault. ");
+            }
+        }
+
+        [Fact]
+        public async Task GetAccessTokenForApp_ReturnsAccessTokenAsync()
+        {
+            // Arrange
+            InitializeTokenAcquisitionObjects();
+
+            Assert.Equal(0, _msalTestTokenCacheProvider.Count);
+
+            // Act
+            string token =
+                await _tokenAcquisition.AcquireTokenForAppAsync(TestConstants.s_scopesForApp).ConfigureAwait(false);
+
+            // Assert
+            Assert.NotNull(token);
+
+            AssertAppTokenInMemoryCache(TestConstants.ConfidentialClientId, 1);
+        }
+
+        [Fact]
+        public async Task GetAccessTokenForApp_WithUserScope_MsalServiceExceptionThrownAsync()
+        {
+             // Arrange
+            InitializeTokenAcquisitionObjects();
+
+            // Act & Assert
+            async Task result() =>
+                await _tokenAcquisition.AcquireTokenForAppAsync(TestConstants.s_scopesForUser).ConfigureAwait(false);
+
+            Exception ex = await Assert.ThrowsAsync<MsalServiceException>(result);
```

Comment:
You can assert the error code as well if you cast directly to the type of exception:

` MsalServiceExceptionex = await Assert.ThrowsAsync<MsalServiceException>(result);`

Reply:
yes..thanks.

---
In reply to: [385619799](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385619799) [](ancestors = 385619799)